calc/l3_route_resolver.go

```diff
@@ -450,10 +411,132 @@ func (c *L3RouteResolver) routesFromBlock(blockKey string, b *model.AllocationBl
 	return routes
 }
 
+// flush() iterates over the CIDRs that are marked dirty in the trie and sends any route updates
+// that it finds.
+func (c *L3RouteResolver) flush() {
+	var buf []ip.V4TrieEntry
+	c.trie.dirtyCIDRs.Iter(func(item interface{}) error {
+		logCxt := logrus.WithField("cidr", item)
+		logCxt.Debug("Flushing dirty route")
+		cidr := item.(ip.V4CIDR)
+
+		// We know the CIDR may be dirty, look up the path through the trie to the CIDR.  This will
+		// give us the information about the enclosing CIDRs.  For example, if we have:
+		// - IP pool     10.0.0.0/16 VXLAN
+		// - IPAM block  10.0.1.0/26 node x
+		// - IP          10.0.0.1/32 node y
+		// Then, we'll see the pool, block and IP in turn on the lookup path allowing us to collect the
+		// relevant information from each.
+		buf = c.trie.t.LookupPath(buf, cidr)
+
+		if len(buf) == 0 {
+			// CIDR is not in the trie.  Nothing to do.  Route removed before it had even been sent?
+			logCxt.Debug("CIDR not in trie, ignoring.")
+			return set.RemoveItem
+		}
+
+		// Otherwise, check if the route is removed.
+		ri := buf[len(buf)-1].Data.(RouteInfo)
+		if ri.WasSent && ri.IsEmpty() {
+			logCxt.Debug("CIDR was sent before but now needs to be removed.")
+			c.callbacks.OnRouteRemove(cidr.String())
+			c.trie.SetRouteSent(cidr, false)
+			return set.RemoveItem
+		}
+
+		rt := &proto.RouteUpdate{
+			Type:       proto.RouteType_CIDR_INFO,
+			IpPoolType: proto.IPPoolType_NONE,
+			Dst:        cidr.String(),
+		}
+		poolAllowsCrossSubnet := false
+		for _, entry := range buf {
+			ri := entry.Data.(RouteInfo)
+			if ri.Pool.Type != proto.IPPoolType_NONE {
+				logCxt.WithField("type", ri.Pool.Type).Debug("Found containing IP pool.")
+				rt.IpPoolType = ri.Pool.Type
+			}
+			if ri.Pool.NATOutgoing {
+				logCxt.Debug("NAT outgoing enabled on this CIDR.")
+				rt.NatOutgoing = true
+			}
+			if ri.Pool.CrossSubnet {
+				logCxt.Debug("Cross-subnet enabled on this CIDR.")
+				poolAllowsCrossSubnet = true
+			}
+			if ri.Block.NodeName != "" {
+				rt.DstNodeName = ri.Block.NodeName
+				rt.DstNodeIP = c.nodeNameToIPAddr[rt.DstNodeName]
+				if rt.DstNodeName == c.myNodeName {
+					logCxt.Debug("Local workload route.")
+					rt.Type = proto.RouteType_LOCAL_WORKLOAD
+				} else {
+					logCxt.Debug("Remote workload route.")
+					rt.Type = proto.RouteType_REMOTE_WORKLOAD
+				}
+			}
+			if ri.Host.NodeName != "" {
+				rt.DstNodeName = ri.Host.NodeName
```

Comment:
Host is the host's own IP rather than a piece of IPAM data.  BPF dataplane needs to know about host IPs as well as workload IPs and we save cycles in the BPF code if we include both in one map.  It's also important for the BPF dataplane to know if the host's IP is _inside_ an IP pool, which sounds crazy but users do that to control NAT outgoing behaviour.

Reply:
I decided to go ahead and set the DstNodeIP in this case too.  I think that'd be what we'd want if it became useful